File: lua/runtime/machine_manager.lua
local machines = require("machines")

local M = {}

function M.new(catalog)
  local self = { catalog = catalog, dirty = false, disabled = {} }

  function self:on_machine_detected(event)
    self.catalog:register(event.machine_type, event.provider, event.machine_id)
    self.disabled[event.machine_id] = nil
    self.dirty = true
  end

  function self:on_machine_removed(event)
    if self.disabled[event.machine_id] then
      return
    end
    self.disabled[event.machine_id] = true
    self.dirty = true
  end

  function self:on_machine_disabled(event)
    self.disabled[event.machine_id] = true
    self.dirty = true
  end

  function self:on_machine_enabled(event)
    self.disabled[event.machine_id] = nil
    self.dirty = true
  end

  function self:build_allocator()
    local active = {}
    for _, machine in ipairs(self.catalog:list_instances()) do
      if not self.disabled[machine.id] then
        table.insert(active, machine)
      end
    end
    self.dirty = false
    return machines.new_allocator(active)
  end

  function self:cleanup_removed()
    for machine_id, _ in pairs(self.disabled) do
      self.catalog:unregister(machine_id)
      self.disabled[machine_id] = nil
    end
    self.dirty = true
  end

  return self
end

return M

File: machine_manager.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Set

from events import EventBus, MachineDetected, MachineDisabled, MachineEnabled, MachineRemoved
from machines import MachineAllocator, MachineCatalog, MachineInstance
from providers.machine.base import MachineProvider


@dataclass
class MachineManager:
    catalog: MachineCatalog
    dirty: bool = False
    _disabled: Set[str] = field(default_factory=set)

    def subscribe(self, bus: EventBus) -> None:
        bus.subscribe(MachineDetected, self.on_machine_detected)
        bus.subscribe(MachineRemoved, self.on_machine_removed)
        bus.subscribe(MachineDisabled, self.on_machine_disabled)
        bus.subscribe(MachineEnabled, self.on_machine_enabled)

    def on_machine_detected(self, event: MachineDetected) -> None:
        if not isinstance(event.provider, MachineProvider):
            raise TypeError("provider must be MachineProvider")
        self.catalog.register(event.machine_type, event.provider, event.machine_id)
        self._disabled.discard(event.machine_id)
        self.dirty = True

    def on_machine_removed(self, event: MachineRemoved) -> None:
        self.catalog.unregister(event.machine_id)
        self._disabled.discard(event.machine_id)
        self.dirty = True

    def on_machine_disabled(self, event: MachineDisabled) -> None:
        self._disabled.add(event.machine_id)
        self.dirty = True

    def on_machine_enabled(self, event: MachineEnabled) -> None:
        self._disabled.discard(event.machine_id)
        self.dirty = True

    def build_allocator(self) -> MachineAllocator:
        active = [m for m in self.catalog.list_instances() if m.id not in self._disabled]
        self.dirty = False
        return MachineAllocator(active)
